service/src/java/org/apache/hive/service/servlet/OTELExporter.java

```diff
@@ -0,0 +1,229 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hive.service.servlet;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.Joiner;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.internal.AttributesMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.hive.ql.QueryDisplay;
+import org.apache.hadoop.hive.ql.QueryInfo;
+import org.apache.hive.service.cli.operation.OperationManager;
+import org.apache.hive.service.cli.session.SessionManager;
+
+public class OTELExporter extends Thread {
+  private static final String INSTRUMENTATION_NAME = OTELExporter.class.getName();
+  private static final Logger LOG = LoggerFactory.getLogger(OTELExporter.class);
+  private final OperationManager operationManager;
+  private Set<String> historicalQueryId;
+  private final long frequency;
+  private final Tracer tracer;
+  private Map<String, Span> queryIdToSpanMap;
+  private Map<String, List<String>> queryIdToTasksMap;
+
+  public OTELExporter(OpenTelemetry openTelemetry, SessionManager sessionManager, long frequency) {
+    this.tracer = openTelemetry.getTracer(INSTRUMENTATION_NAME);
+    this.operationManager = sessionManager.getOperationManager();
+    this.historicalQueryId = new HashSet<>();
+    this.frequency = frequency;
+    this.queryIdToSpanMap = new HashMap<>();
+    this.queryIdToTasksMap = new HashMap<>();
+  }
+
+  @Override
+  public void run() {
+    while (true) {
+      exposeMetricsToOTEL();
+      try {
+        Thread.sleep(frequency);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  public void exposeMetricsToOTEL() {
+    List<QueryInfo> liveQueries = operationManager.getLiveQueryInfos();
+    List<QueryInfo> historicalQueries = operationManager.getHistoricalQueryInfos();
+
+    LOG.debug("Found {} liveQueries and {} historicalQueries", liveQueries.size(), historicalQueries.size());
+
+    for (QueryInfo lQuery: liveQueries){
+
+      while (lQuery.getQueryDisplay() == null) {
+        try {
+          Thread.sleep(500);
```

Comment:
Why is there are sleep here?

Reply:
In some cases, OTEL service was capturing the liveQueries in a state where queryDisplay was not initialised and leading to Null Pointer Exception. This is changed to drop such queries for the current OTEL run and capture it in next run.